// galeforce/dist/galeforce/actions/mixins/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TakesCard = exports.TakesLorRegion = exports.TakesLorSet = exports.TakesVersion = exports.TakesSpell = exports.TakesSkin = exports.TakesLocale = exports.TakesAssetPath = exports.TakesDataDragonId = exports.TakesChampion = exports.TakesTournamentId = exports.TakesTournamentCode = exports.TakesTeamId = exports.TakesSummonerName = exports.TakesSummonerId = exports.TakesRiotId = exports.TakesRegion = exports.TakesTier = exports.TakesQueue = exports.TakesQuery = exports.TakesPUUID = exports.TakesMatchId = exports.TakesLeagueId = exports.TakesGame = exports.TakesDivision = exports.TakesChampionId = exports.TakesBody = exports.TakesActId = exports.TakesAccountId = void 0;
var account_id_1 = require("./account-id");
Object.defineProperty(exports, "TakesAccountId", { enumerable: true, get: function () { return account_id_1.TakesAccountId; } });
var act_id_1 = require("./act-id");
Object.defineProperty(exports, "TakesActId", { enumerable: true, get: function () { return act_id_1.TakesActId; } });
var body_1 = require("./body");
Object.defineProperty(exports, "TakesBody", { enumerable: true, get: function () { return body_1.TakesBody; } });
var champion_id_1 = require("./champion-id");
Object.defineProperty(exports, "TakesChampionId", { enumerable: true, get: function () { return champion_id_1.TakesChampionId; } });
var division_1 = require("./division");
Object.defineProperty(exports, "TakesDivision", { enumerable: true, get: function () { return division_1.TakesDivision; } });
var game_1 = require("./game");
Object.defineProperty(exports, "TakesGame", { enumerable: true, get: function () { return game_1.TakesGame; } });
var league_id_1 = require("./league-id");
Object.defineProperty(exports, "TakesLeagueId", { enumerable: true, get: function () { return league_id_1.TakesLeagueId; } });
var match_id_1 = require("./match-id");
Object.defineProperty(exports, "TakesMatchId", { enumerable: true, get: function () { return match_id_1.TakesMatchId; } });
var puuid_1 = require("./puuid");
Object.defineProperty(exports, "TakesPUUID", { enumerable: true, get: function () { return puuid_1.TakesPUUID; } });
var query_1 = require("./query");
Object.defineProperty(exports, "TakesQuery", { enumerable: true, get: function () { return query_1.TakesQuery; } });
var queue_1 = require("./queue");
Object.defineProperty(exports, "TakesQueue", { enumerable: true, get: function () { return queue_1.TakesQueue; } });
var tier_1 = require("./tier");
Object.defineProperty(exports, "TakesTier", { enumerable: true, get: function () { return tier_1.TakesTier; } });
var region_1 = require("./region");
Object.defineProperty(exports, "TakesRegion", { enumerable: true, get: function () { return region_1.TakesRegion; } });
var riot_id_1 = require("./riot-id");
Object.defineProperty(exports, "TakesRiotId", { enumerable: true, get: function () { return riot_id_1.TakesRiotId; } });
var summoner_id_1 = require("./summoner-id");
Object.defineProperty(exports, "TakesSummonerId", { enumerable: true, get: function () { return summoner_id_1.TakesSummonerId; } });
var summoner_name_1 = require("./summoner-name");
Object.defineProperty(exports, "TakesSummonerName", { enumerable: true, get: function () { return summoner_name_1.TakesSummonerName; } });
var team_id_1 = require("./team-id");
Object.defineProperty(exports, "TakesTeamId", { enumerable: true, get: function () { return team_id_1.TakesTeamId; } });
var tournament_code_1 = require("./tournament-code");
Object.defineProperty(exports, "TakesTournamentCode", { enumerable: true, get: function () { return tournament_code_1.TakesTournamentCode; } });
var tournament_id_1 = require("./tournament-id");
Object.defineProperty(exports, "TakesTournamentId", { enumerable: true, get: function () { return tournament_id_1.TakesTournamentId; } });
var champion_1 = require("./champion");
Object.defineProperty(exports, "TakesChampion", { enumerable: true, get: function () { return champion_1.TakesChampion; } });
var data_dragon_id_1 = require("./data-dragon-id");
Object.defineProperty(exports, "TakesDataDragonId", { enumerable: true, get: function () { return data_dragon_id_1.TakesDataDragonId; } });
var asset_path_1 = require("./asset-path");
Object.defineProperty(exports, "TakesAssetPath", { enumerable: true, get: function () { return asset_path_1.TakesAssetPath; } });
var locale_1 = require("./locale");
Object.defineProperty(exports, "TakesLocale", { enumerable: true, get: function () { return locale_1.TakesLocale; } });
var skin_1 = require("./skin");
Object.defineProperty(exports, "TakesSkin", { enumerable: true, get: function () { return skin_1.TakesSkin; } });
var spell_1 = require("./spell");
Object.defineProperty(exports, "TakesSpell", { enumerable: true, get: function () { return spell_1.TakesSpell; } });
var version_1 = require("./version");
Object.defineProperty(exports, "TakesVersion", { enumerable: true, get: function () { return version_1.TakesVersion; } });
var lor_set_1 = require("./lor-set");
Object.defineProperty(exports, "TakesLorSet", { enumerable: true, get: function () { return lor_set_1.TakesLorSet; } });
var lor_region_1 = require("./lor-region");
Object.defineProperty(exports, "TakesLorRegion", { enumerable: true, get: function () { return lor_region_1.TakesLorRegion; } });
var card_1 = require("./card");
Object.defineProperty(exports, "TakesCard", { enumerable: true, get: function () { return card_1.TakesCard; } });
