// galeforce/dist/galeforce/interfaces/dto/riot-api/platform/platform-data.js
"use strict";
/* eslint-disable camelcase */
/*
    The PlatformDataDTO is an interface for status data returned from V4 endpoints.
*/
Object.defineProperty(exports, "__esModule", { value: true });
var MaintenanceStatus;
(function (MaintenanceStatus) {
    MaintenanceStatus["SCHEDULED"] = "scheduled";
    MaintenanceStatus["IN_PROGRESS"] = "in_progress";
    MaintenanceStatus["COMPLETE"] = "complete";
})(MaintenanceStatus || (MaintenanceStatus = {}));
var IncidentSeverity;
(function (IncidentSeverity) {
    IncidentSeverity["INFO"] = "info";
    IncidentSeverity["WARNING"] = "warning";
    IncidentSeverity["CRITICAL"] = "critical";
})(IncidentSeverity || (IncidentSeverity = {}));
var Platform;
(function (Platform) {
    Platform["WINDOWS"] = "windows";
    Platform["MACOS"] = "macos";
    Platform["ANDROID"] = "android";
    Platform["IOS"] = "ios";
    Platform["PS4"] = "ps4";
    Platform["XBONE"] = "xbone";
    Platform["SWITCH"] = "switch";
})(Platform || (Platform = {}));
var PublishLocations;
(function (PublishLocations) {
    PublishLocations["RIOT_CLIENT"] = "riotclient";
    PublishLocations["RIOT_STATUS"] = "riotstatus";
    PublishLocations["GAME"] = "game";
})(PublishLocations || (PublishLocations = {}));
