// galeforce/dist/riot-api/enums/regions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LorRegion = exports.DataDragonRegion = exports.ValorantRegion = exports.RiotRegion = exports.LeagueRegion = void 0;
var LeagueRegion;
(function (LeagueRegion) {
    LeagueRegion["BRAZIL"] = "br1";
    LeagueRegion["EUROPE_NORTHEAST"] = "eun1";
    LeagueRegion["EUROPE_WEST"] = "euw1";
    LeagueRegion["KOREA"] = "kr";
    LeagueRegion["LATIN_AMERICA_NORTH"] = "la1";
    LeagueRegion["LATIN_AMERICA_SOUTH"] = "la2";
    LeagueRegion["NORTH_AMERICA"] = "na1";
    LeagueRegion["OCEANIA"] = "oc1";
    LeagueRegion["RUSSIA"] = "ru";
    LeagueRegion["TURKEY"] = "tr1";
    LeagueRegion["JAPAN"] = "jp1";
})(LeagueRegion || (LeagueRegion = {}));
exports.LeagueRegion = LeagueRegion;
var RiotRegion;
(function (RiotRegion) {
    RiotRegion["AMERICAS"] = "americas";
    RiotRegion["ASIA"] = "asia";
    RiotRegion["EUROPE"] = "europe";
    RiotRegion["ESPORTS"] = "esports";
})(RiotRegion || (RiotRegion = {}));
exports.RiotRegion = RiotRegion;
var LorRegion;
(function (LorRegion) {
    LorRegion["AMERICAS"] = "americas";
    LorRegion["ASIA"] = "asia";
    LorRegion["EUROPE"] = "europe";
    LorRegion["SOUTHEAST_ASIA"] = "sea";
})(LorRegion || (LorRegion = {}));
exports.LorRegion = LorRegion;
var ValorantRegion;
(function (ValorantRegion) {
    ValorantRegion["ASIA_PACIFIC"] = "ap";
    ValorantRegion["BRAZIL"] = "br";
    ValorantRegion["EUROPE"] = "eu";
    ValorantRegion["KOREA"] = "kr";
    ValorantRegion["LATIN_AMERICA"] = "latam";
    ValorantRegion["NORTH_AMERICA"] = "na";
    ValorantRegion["PBE"] = "pbe1";
    ValorantRegion["ESPORTS"] = "esports";
})(ValorantRegion || (ValorantRegion = {}));
exports.ValorantRegion = ValorantRegion;
var DataDragonRegion;
(function (DataDragonRegion) {
    DataDragonRegion["BRAZIL"] = "br";
    DataDragonRegion["EUROPE_NORTHEAST"] = "eune";
    DataDragonRegion["EUROPE_WEST"] = "euw";
    DataDragonRegion["KOREA"] = "kr";
    DataDragonRegion["LATIN_AMERICA_NORTH"] = "lan";
    DataDragonRegion["LATIN_AMERICA_SOUTH"] = "las";
    DataDragonRegion["NORTH_AMERICA"] = "na";
    DataDragonRegion["OCEANIA"] = "oce";
    DataDragonRegion["RUSSIA"] = "ru";
    DataDragonRegion["TURKEY"] = "tr";
    DataDragonRegion["JAPAN"] = "jp";
})(DataDragonRegion || (DataDragonRegion = {}));
exports.DataDragonRegion = DataDragonRegion;
